jdaviz/app.py

```diff
@@ -106,48 +107,67 @@ def to_unit(self, data, cid, values, original_units, target_units):
-                                    lambda x: (x / spec.meta['_pixel_scale_factor']),
-                                    lambda x: x)]
-                        else:
-                            # Surface Brightness -> Flux
-                            eqv = u.spectral_density(spec.spectral_axis)
+                eqv = []
+                # if spectrum data collection item is in Surface Brightness units
+                if u.sr in spec.unit.bases and len(values) != 2:
+                    # Data item in data collection does not update from conversion/translation.
+                    # App wide orginal data units are used for conversion, orginal_units and
+                    # target_units dicate the conversion to take place.
+                    if (u.sr in u.Unit(original_units).bases) and \
+                       (u.sr not in u.Unit(target_units).bases):
+                        # Surface Brightness -> Flux
+                        eqv = [(u.MJy / u.sr,
+                                u.MJy,
+                                lambda x: (x * np.array(spec.meta['_pixel_scale_factor'])),
+                                lambda x: x)]
+                    else:
+                        # Flux -> Surface Brightness
+                        eqv = u.spectral_density(spec.spectral_axis)
+                # if spectrum data collection item is in Flux units
+                elif u.sr not in spec.unit.bases and len(values) != 2:
+                    # Data item in data collection does not update from conversion/translation.
+                    # App wide orginal data units are used for conversion, orginal_units and
+                    # target_units dicate the conversion to take place.
+                    if (u.sr not in u.Unit(original_units).bases) and \
+                       (u.sr in u.Unit(target_units).bases):
+                        # Flux -> Surface Brightness
+                        eqv = [(u.MJy,
+                                u.MJy / u.sr,
+                                lambda x: (x / np.array(spec.meta['_pixel_scale_factor'])),
+                                lambda x: x)]
 
+                    else:
+                        # Surface Brightness -> Flux
+                        eqv = u.spectral_density(spec.spectral_axis)
                 elif len(values) == 2:
                     # Need this for setting the y-limits
                     spec_limits = [spec.spectral_axis[0].value, spec.spectral_axis[-1].value]
                     eqv = u.spectral_density(spec_limits * spec.spectral_axis.unit)
 
+                    if '_pixel_scale_factor' in spec.meta:
+                        # get min and max scale factors, to use with min and max of spec for
+                        # y-limits. Make sure they are Quantities (can be numpy.float64).
+                        pixel_scale_min = (Quantity(spec.meta['_pixel_scale_factor'][0])).value
+                        pixel_scale_max = (Quantity(spec.meta['_pixel_scale_factor'][-1])).value
```

Comment:
Does anything guarantee that `spec.meta['_pixel_scale_factor']` is sorted? If not, use `min` and `max` rather than `[0]` and `[-1]`.

Reply:
I was following the convention used for getting the limits of the spectrum axis, and do agree ensuring that these 2 values used are actually the min and max of the scale factor array!